Add support for sym_ite

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111440

This PR supports sym_ite. This is useful for converting SymBool to SymInt in e.g. #109916. Internally, it uses sympy.Piecewise. We cannot use sympy.ITE because it expects the arguments and output all to be boolean type but we want return SymInt type when converting a SymBool to SymInt. So we use sympy.Piecewise to denote the symbolic relationship. 

Note that this pr uses the range analysis for sympy.Piecewise implemented in https://github.com/pytorch/pytorch/blob/main/torch/utils/_sympy/value_ranges.py. 

Test Plan:
See added test.

ydwu4 committed:
Add support for sym_ite

[ghstack-poisoned]

ydwu4 committed:
Update on "Add support for sym_ite"




[ghstack-poisoned]

ydwu4 committed:
Update on "Add support for sym_ite"




[ghstack-poisoned]

ydwu4 committed:
Update on "Add support for sym_ite"


This PR supports sym_ite. This is useful for converting SymBool to SymInt in e.g. #109916. 

Test Plan:
See added test.


[ghstack-poisoned]

ydwu4 committed:
Update on "Add support for sym_ite"


This PR supports sym_ite. This is useful for converting SymBool to SymInt in e.g. #109916. Internally, it uses sympy.Piecewise. We cannot use sympy.ITE because it expects the arguments and output all to be boolean type but we want return SymInt type when converting a SymBool to SymInt. So we use sympy.Piecewise to denote the symbolic relationship. 

Note that this pr uses the range analysis for sympy.Piecewise implemented in https://github.com/pytorch/pytorch/blob/main/torch/utils/_sympy/value_ranges.py. 

Test Plan:
See added test.


[ghstack-poisoned]

ydwu4 committed:
Update on "Add support for sym_ite"


This PR supports sym_ite. This is useful for converting SymBool to SymInt in e.g. #109916. Internally, it uses sympy.Piecewise. We cannot use sympy.ITE because it expects the arguments and output all to be boolean type but we want return SymInt type when converting a SymBool to SymInt. So we use sympy.Piecewise to denote the symbolic relationship. 

Note that this pr uses the range analysis for sympy.Piecewise implemented in https://github.com/pytorch/pytorch/blob/main/torch/utils/_sympy/value_ranges.py. 

Test Plan:
See added test.


[ghstack-poisoned]

ydwu4 committed:
Update on "Add support for sym_ite"


This PR supports sym_ite. This is useful for converting SymBool to SymInt in e.g. #109916. Internally, it uses sympy.Piecewise. We cannot use sympy.ITE because it expects the arguments and output all to be boolean type but we want return SymInt type when converting a SymBool to SymInt. So we use sympy.Piecewise to denote the symbolic relationship. 

Note that this pr uses the range analysis for sympy.Piecewise implemented in https://github.com/pytorch/pytorch/blob/main/torch/utils/_sympy/value_ranges.py. 

Test Plan:
See added test.


[ghstack-poisoned]

ezyang commented on test/test_dynamic_shapes.py:
```diff
@@ -423,6 +423,47 @@ def test_sym_ceil(self):
+        self.assertTrue(r2 is f)
+        b3 = t == 5
+        r3 = torch.fx.experimental.symbolic_shapes.sym_ite(b3, t, f)
+        self.assertEqual(r3, 5)
```
This test phrasing is a bit dorky because this is generating the guard. You want some test before this assertEqual showing that there is no guard created just because you sym_ite an expression.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1344,6 +1344,44 @@ def cast_symbool_to_symint_guardless(symbool: torch.SymBool) -> torch.SymInt:
+        if pred.node.is_constant():
+            constant_pred = guard_bool(pred)
+            return then_val if constant_pred else else_val
+        return _sym_ite(pred, then_val, else_val)
```
Check how the other sym ops go, they have to go through a dynamic dispatch on SymNode first. That allows for C++ to override the meaning of operations like this. As implemented today, this will only work if you get a Python SymNode.

ezyang replied:
This also implies that you'll have to add a new virtual method to SymNodeImpl in C++

ydwu4 replied:
Is sym_max a good example to follow? How can I test the c++ implementation?

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1344,6 +1344,44 @@ def cast_symbool_to_symint_guardless(symbool: torch.SymBool) -> torch.SymInt:
+        raise NotImplementedError(f"sym_ite only supports bool or SymBool pred but got {type(pred)}.")
+
+def _sym_ite(pred, then_val, else_val):
+    pred_node = pred.node
```
Once you go through C++, it will no longer be necessary to convert these to nodes

ydwu4 replied:
Why is that? Do the python implementation of sym ops call the c++ implementation? Can I have a code pointer?

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1344,6 +1344,44 @@ def cast_symbool_to_symint_guardless(symbool: torch.SymBool) -> torch.SymInt:
+        return SymNode(out, pred_node.shape_env, then_node.pytype, out_hint, fx_node=fx_node)
+
+    ret = wrap_node(_sym_node_ite(pred_node, then_node, else_node))
+    return get_constant(ret) if ret.node.is_constant() else ret
```
and you wouldn't need this code either

ydwu4 committed:
Update on "Add support for sym_ite"


This PR supports sym_ite. This is useful for converting SymBool to SymInt in e.g. #109916. Internally, it uses sympy.Piecewise. We cannot use sympy.ITE because it expects the arguments and output all to be boolean type but we want return SymInt type when converting a SymBool to SymInt. So we use sympy.Piecewise to denote the symbolic relationship. 

Note that this pr uses the range analysis for sympy.Piecewise implemented in https://github.com/pytorch/pytorch/blob/main/torch/utils/_sympy/value_ranges.py. 

Test Plan:
See added test.


[ghstack-poisoned]

ydwu4 committed:
Update on "Add support for sym_ite"


This PR supports sym_ite. This is useful for converting SymBool to SymInt in e.g. #109916. Internally, it uses sympy.Piecewise. We cannot use sympy.ITE because it expects the arguments and output all to be boolean type but we want return SymInt type when converting a SymBool to SymInt. So we use sympy.Piecewise to denote the symbolic relationship. 

Note that this pr uses the range analysis for sympy.Piecewise implemented in https://github.com/pytorch/pytorch/blob/main/torch/utils/_sympy/value_ranges.py. 

Test Plan:
See added test.


[ghstack-poisoned]

ezyang reviewed (approved):
Thanks for taking the time to do the C++ impl, I appreciate it.

ydwu4 commented:
@pytorchbot merge
